decryptor/postgresql/data_encoder.go

```diff
@@ -247,3 +231,101 @@ func (p *PgSQLDataDecoderProcessor) OnColumn(ctx context.Context, data []byte) (
 	}
 	return p.decodeText(ctx, data, columnSetting, columnInfo, logger)
 }
+
+// encodingValue represents a (possibly parsed and prepared) value that is
+// ready to be encoded
+type encodingValue interface {
+	asBinary() []byte
+	asText() []byte
+}
+
+// byteSequenceValue is an abstraction over all byte-sequence values -- strings
+// and []byte (because they are encoded in the same way)
+type byteSequenceValue struct {
+	seq []byte
+}
+
+func (v *byteSequenceValue) asBinary() []byte { return v.seq }
+func (v *byteSequenceValue) asText() []byte {
+	if utils.IsPrintablePostgresqlString(v.seq) {
+		return v.seq
+	}
+	return utils.PgEncodeToHex(v.seq)
+}
+
+// intValue represents a {size*8}-bit integer ready for encoding
+type intValue struct {
+	size     int
+	value    int64
+	strValue string
+}
+
+func (v *intValue) asBinary() []byte {
+	newData := make([]byte, v.size)
+	switch v.size {
+	case 4:
+		binary.BigEndian.PutUint32(newData, uint32(v.value))
+	case 8:
+		binary.BigEndian.PutUint64(newData, uint64(v.value))
+	}
+	return newData
+}
+
+func (v *intValue) asText() []byte {
+	return []byte(v.strValue)
+}
+
+// encodeDefault returns wrapped default value from settings ready for encdoing
+// returns nil if something went wrong, which in many cases indicates that the
+// original value should be returned as it is
+func encodeDefault(setting config.ColumnEncryptionSetting, logger *logrus.Entry) encodingValue {
+	strValue := setting.GetDefaultDataValue()
+	if strValue == nil {
+		logger.Errorln("Default value is not specified")
+		return nil
+	}
+
+	dataType := setting.GetEncryptedDataType()
+
+	switch dataType {
+	case common2.EncryptedType_String:
+		return &byteSequenceValue{seq: []byte(*strValue)}
+	case common2.EncryptedType_Bytes:
+		binValue, err := base64.StdEncoding.DecodeString(*strValue)
+		if err != nil {
+			logger.WithError(err).Errorln("Can't decode base64 default value")
+			return nil
+		}
+		return &byteSequenceValue{seq: binValue}
+	case common2.EncryptedType_Int32, common2.EncryptedType_Int64:
+		size := 8
+		if dataType == common2.EncryptedType_Int32 {
+			size = 4
+		}
+		value, err := strconv.ParseInt(*strValue, 10, 64)
+		if err != nil {
+			logger.WithError(err).Errorln("Can't parse default integer value")
+			return nil
+		}
+
+		return &intValue{size: size, value: value, strValue: *strValue}
+	}
+	return nil
+}
+
+// onFail returns either an error, which should be returned, or value, which
+// should be encoded, because there is some problem with original, or `nil`
+// which indicates that original value should be returned as is.
+func onFail(setting config.ColumnEncryptionSetting, logger *logrus.Entry) (encodingValue, error) {
```

Comment:
maybe name it like `encodeOnFail`? because not obvious what function does ...
Plus, what about third option on fail: return ciphertext as is? as we discussed, first option was return as is even with incorrect data type (our first version). Second option was to return default value. And third is return error.
So, to return ciphertext as is we should accept here `data` parameter too.

Where it may be useful return ciphertext. For example for apps that can decrypt it by itself. If app has access to keys or it can use AcraTranslator for failure cases.